### PR TITLE
CSharp Refactor - Instance Config Endpoint

### PR DESCRIPTION
**Pull request by nharper285**
## Summary of the Pull Request

_What is this about?_
Working through refactor based-off issue #1747 
## PR Checklist
* [ ] Applies to work item: #xxx
* [ ] CLA signed. If not, go over [here](https://cla.opensource.microsoft.com/microsoft/onefuzz) and sign the CLI.
* [ ] Tests added/passed
* [ ] Requires documentation to be updated
* [ ] I've discussed this with core contributors already. If not checked, I'm ready to accept this work might be rejected in favor of a different grand plan. Issue number where discussion took place: #xxx

## Info on Pull Request

_What does this include?_

## Validation Steps Performed

_How does someone test & validate?_

### CODE REVIEW

File: src/ApiService/ApiService/Functions/InstanceConfig.cs [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+        var request = await RequestHandling.ParseRequest<InstanceConfigUpdate>(req);
+
+        if (!request.IsOk) {
+            return await _context.RequestHandling.NotOk(
+                req,
+                request.ErrorV,
+                context: "instance_config update");
+        }
+        var config = await _context.ConfigOperations.Fetch();
+        var answer = await _auth.CheckRequireAdmins(req);
```
Comment (Porges) [inline src/ApiService/ApiService/Functions/InstanceConfig.cs:47]: You can await both to do these operations in parallel:

(also needs `using System.Threading.Tasks;` at top) 

```suggestion
        var (config, answer) = await (
            _context.ConfigOperations.Fetch(),
            _auth.CheckRequireAdmins(req));
```

File: src/ApiService/ApiService/Functions/InstanceConfig.cs [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+            return await _context.RequestHandling.NotOk(req, answer.ErrorV, "instance_config update");
+        }
+        var updateNsg = false;
+        if (request.OkV.config.ProxyNsgConfig is not null && config.ProxyNsgConfig is not null) {
+            var requestConfig = request.OkV.config.ProxyNsgConfig;
+            var currentConfig = config.ProxyNsgConfig;
+            if (!((new HashSet<string>(requestConfig.AllowedServiceTags)).SetEquals(new HashSet<string>(currentConfig.AllowedServiceTags))) || !((new HashSet<string>(requestConfig.AllowedIps)).SetEquals(new HashSet<string>(currentConfig.AllowedIps)))) {
+                updateNsg = true;
+            }
+        }
```
Comment (Porges) [inline src/ApiService/ApiService/Functions/InstanceConfig.cs:58]: Suggest pattern-matching into some variables, and the comparison can be simplified a little:

```suggestion
        if (request.OkV.config.ProxyNsgConfig is NetworkSecurityGroupConfig requestConfig
            && config.ProxyNsgConfig is NetworkSecurityGroupConfig currentConfig) {
            if (!requestConfig.AllowedServiceTags.ToHashSet().SetEquals(currentConfig.AllowedServiceTags)
                || !requestConfig.AllowedIps.ToHashSet().SetEquals(currentConfig.AllowedIps)) {
                updateNsg = true;
            }
        }
```

File: src/ApiService/ApiService/Functions/InstanceConfig.cs [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+                req,
+                request.ErrorV,
+                context: "instance_config update");
+        }
+        var config = await _context.ConfigOperations.Fetch();
+        var answer = await _auth.CheckRequireAdmins(req);
+        if (!answer.IsOk) {
+            return await _context.RequestHandling.NotOk(req, answer.ErrorV, "instance_config update");
+        }
+        var updateNsg = false;
```
Comment (Porges) [inline src/ApiService/ApiService/Functions/InstanceConfig.cs:51]: `updateNsg` should probably also be true if there is a request config and not a stored config

File: src/ApiService/ApiService/Functions/InstanceConfig.cs [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+            var requestConfig = request.OkV.config.ProxyNsgConfig;
+            var currentConfig = config.ProxyNsgConfig;
+            if (!((new HashSet<string>(requestConfig.AllowedServiceTags)).SetEquals(new HashSet<string>(currentConfig.AllowedServiceTags))) || !((new HashSet<string>(requestConfig.AllowedIps)).SetEquals(new HashSet<string>(currentConfig.AllowedIps)))) {
+                updateNsg = true;
+            }
+        }
+        await _context.ConfigOperations.Save(request.OkV.config, false, false);
+        if (updateNsg) {
+            await foreach (var nsg in _context.NsgOperations.ListNsgs()) {
+                _log.Info($"Checking if nsg: {nsg.Data.Location!} ({nsg.Data.Name}) owned by OneFuz");
```
Comment (Porges) [inline src/ApiService/ApiService/Functions/InstanceConfig.cs:62]:
```suggestion
                _log.Info($"Checking if nsg: {nsg.Data.Location!} ({nsg.Data.Name}) owned by OneFuzz");
```

**Review by Porges — APPROVED** [review]

**Review by stishkin — APPROVED** [review]

File: src/ApiService/ApiService/Functions/InstanceConfig.cs (hunk 3 above)
Comment (nharper285) [reply src/ApiService/ApiService/Functions/InstanceConfig.cs:51]: Stored config should never be null!

File: src/ApiService/ApiService/Functions/InstanceConfig.cs (hunk 2 above)
Comment (nharper285) [reply src/ApiService/ApiService/Functions/InstanceConfig.cs:58]: Oh this is nice. Thanks for the suggestion.

**Review by tevoinea — APPROVED** [review]

**Pull request merged**